.emacs.d/lisp/visual/line-numbers-c.el

```diff
@@ -0,0 +1,44 @@
+(defun kotct/line-numbers--set-up-linum ()
+  "Sets up `linum-mode' to run globally except in a few cases."
+  (progn
+    ;; enable linum globally
+    (global-linum-mode 1)
+
+    ;; use a lambda to generate the linum formatting
+    ;; this will use the minimum number of columns,
+    ;; but also always right-align
+    (setq linum-format (lambda (line)
+                         (let* ((w (length (number-to-string (count-lines (point-min) (point-max)))))
+                                (thing (concat " %" (number-to-string (max 2 w)) "d ")))
+                           (propertize (format thing line) 'face 'linum))))
+
+    ;; disable linum in certain modes
+    (require 'linum-off)
+    (add-to-list 'linum-disabled-modes-list 'package-menu-mode)
+    (add-to-list 'linum-disabled-modes-list 'magit--mode)))
+
+(defun kotct/display-line-numbers--turn-off ()
+  "Turns off `display-line-numbers-mode' for the current buffer."
+  (display-line-numbers-mode -1))
+
+(defun kotct/line-numbers--set-up-display-line-numbers ()
+  "Sets up `display-line-numbers-mode'.
+
+This mode was added in Emacs 26.1 and is recommended over `linum-mode' because
+it generally works better."
+  (progn
+    ;; inhibit global-linum-mode
+    (global-linum-mode -1)
+    (global-display-line-numbers-mode 1)
+
+    ;; add some hooks to turn off display-line-numbers mode
+    (add-hook 'dired-mode-hook #'kotct/display-line-numbers--turn-off)
+    (add-hook 'git-commit-mode-hook #'kotct/display-line-numbers--turn-off)
+    (add-hook 'magit-mode-hook #'kotct/display-line-numbers--turn-off)
+    (add-hook 'package-menu-mode-hook #'kotct/display-line-numbers--turn-off)))
```

Comment:
A global function seems a little overkill for this simple thing, but idk. I might let-bind a lambda and use that instead, or dolist over the hooks and put the lambda inline (as in
```elisp
(dolist (hook '(dired-mode-hook ...))
  (add-hook hook (lambda () (display-line-numbers-mode -1))))
```
) but this seems fine. Just because it's not how I would do it doesn't make it wrong.

Reply:
I'm just thinking if we end up wanting to toggle…

My big hangup on adding lambdas over functions is that I feel like this makes it easier to potentially _remove_ hooks, whereas if you set lambdas it's much harder to take care of.  Since the lambdas are doing something, and that something is something that has a pretty descriptive name, I feel like this is the best approach. (I'd obviously not do this if we had special handling such as relative line numbering for `git-commit-mode` or something ridiculous like that.)

Reply:
Actually, I do have an idea.  Couldn't we just do `(add-hook hook #'kotct/turn-off-line-numbers)` or whatever that is inside that enclosure?  If so, I think that's more elegant.  (More DRY.)

Reply:
>this makes it easier to potentially remove hooks, whereas if you set lambdas it's much harder to take care of

This is a good point actually. I hadn't thought of this.